### ISSUE TITLE
500 Internal server error 

### ISSUE THREAD

**Issue by kotjarka** [gramps-project/Gramps.js#208]
I observed strange behavior. I added a family as usual, two people and a child. After that, when tried to enter the family list (/families), the server reports 500 Internal server error . 
I can still display this problematic family from history through the link /family/F0021. But this one family is not visible on the /tree. 
I do not know what went wrong, I suspect that something was incorrectly saved in the database. Can I fix it using DB Browser for SQLite?
Gramps-web is installed using docker-compose.yml via podman-compose.

**Comment by DavidMStraub** [comment]
Hi, can you please post your server logs? `docker-compose logs --tail 100 grampsweb`

**Comment by kotjarka** [comment]
Don't have docker-compose installed, I`m running containers with podman. Log from podman instead:

```
$ podman logs --tail 100 gramps_grampsweb_1
  File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1263, in _get_from_handle
    raise HandleError('Handle is None')
gramps.gen.errors.HandleError: Handle is None
[2023-04-30 18:51:38 +0000] [9] [ERROR] Error handling request /api/families/?locale=pl&profile=self&keys=gramps_id,profile,change&page=1&pagesize=20
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 136, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 179, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2551, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2531, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/auth.py", line 44, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/flask/views.py", line 107, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)
  File "/usr/local/lib/python3.9/dist-packages/flask/views.py", line 188, in dispatch_request
    return current_app.ensure_sync(meth)(**kwargs)
  File "/usr/local/lib/python3.9/dist-packages/webargs/core.py", line 594, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/base.py", line 412, in get
    [
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/base.py", line 413, in <listcomp>
    self.full_object(query_method(handle), args, locale=locale)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/base.py", line 85, in full_object
    obj = self.object_extend(obj, args, locale=locale)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/families.py", line 55, in object_extend
    obj.profile = get_family_profile_for_object(
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/util.py", line 482, in get_family_profile_for_object
    divorce, divorce_event = get_divorce_profile(
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/util.py", line 314, in get_divorce_profile
    event = get_divorce_or_fallback(db_handle, family)
  File "/usr/lib/python3/dist-packages/gramps/gen/utils/db.py", line 231, in get_divorce_or_fallback
    divorce_ref = get_event_ref(db, family, EventType.DIVORCE)
  File "/usr/lib/python3/dist-packages/gramps/gen/utils/db.py", line 178, in get_event_ref
    event = db.get_event_from_handle(event_ref.ref)
  File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1273, in get_event_from_handle
    return self._get_from_handle(EVENT_KEY, Event, handle)
  File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1263, in _get_from_handle
    raise HandleError('Handle is None')
gramps.gen.errors.HandleError: Handle is None
[2023-04-30 18:51:38 +0000] [17] [ERROR] Error handling request /api/families/?locale=pl&profile=self&keys=gramps_id,profile,change&page=1&pagesize=20
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 136, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/local/lib/python3.9/dist-packages/gunicorn/workers/sync.py", line 179, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2551, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2531, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 2528, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1825, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1823, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.9/dist-packages/flask/app.py", line 1799, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/auth.py", line 44, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/flask/views.py", line 107, in view
    return current_app.ensure_sync(self.dispatch_request)(**kwargs)
  File "/usr/local/lib/python3.9/dist-packages/flask/views.py", line 188, in dispatch_request
    return current_app.ensure_sync(meth)(**kwargs)
  File "/usr/local/lib/python3.9/dist-packages/webargs/core.py", line 594, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/base.py", line 412, in get
    [
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/base.py", line 413, in <listcomp>
    self.full_object(query_method(handle), args, locale=locale)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/base.py", line 85, in full_object
    obj = self.object_extend(obj, args, locale=locale)
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/families.py", line 55, in object_extend
    obj.profile = get_family_profile_for_object(
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/util.py", line 482, in get_family_profile_for_object
    divorce, divorce_event = get_divorce_profile(
  File "/usr/local/lib/python3.9/dist-packages/gramps_webapi/api/resources/util.py", line 314, in get_divorce_profile
    event = get_divorce_or_fallback(db_handle, family)
  File "/usr/lib/python3/dist-packages/gramps/gen/utils/db.py", line 231, in get_divorce_or_fallback
    divorce_ref = get_event_ref(db, family, EventType.DIVORCE)
  File "/usr/lib/python3/dist-packages/gramps/gen/utils/db.py", line 178, in get_event_ref
    event = db.get_event_from_handle(event_ref.ref)
  File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1273, in get_event_from_handle
    return self._get_from_handle(EVENT_KEY, Event, handle)
  File "/usr/lib/python3/dist-packages/gramps/gen/db/generic.py", line 1263, in _get_from_handle
    raise HandleError('Handle is None')
gramps.gen.errors.HandleError: Handle is None
ERROR:root:Error parsing list of recent DBs from file /root/.gramps/recent-files-gramps.xml: no element found: line 1, column 0.
This might indicate a damage to your files.
If you're sure there is no problem with other files, delete it, and restart Gramps.
```

**Comment by DavidMStraub** [comment]
Ok thanks, do you remember the exact steps you used in the UI to create and link the family and persons?

**Comment by DavidMStraub** [comment]
In the meantime, regarding your question 

> Can I fix it using DB Browser for SQLite?

Can you please check the `family` table for the entry with Gramps ID `F0021`? 

Also, it looks like the error is caused by a divorce event with empty handle, can you please check the `event` table for the rows with the most recent (highest) `change` timestamp?

Thanks!

**Comment by kotjarka** [comment]
Thank for your response. 
Family was added as usual, "+" button on top menu and then in form (/new_family) I selected father, mother and 1 child (all 3 persons added immediately before).
In `family` table in database, `F0021` record exists and look like others (don't know how to inspect `blob` field.)
There is no divorce event in database. Last rows in `event` table are 2x birth and 1x death (from members of family `F0021`). 
I also tried export and check in desktop Gramps, but without succes (Gramps XML option returns empty index.html file)

**Comment by DavidMStraub** [comment]
>  I also tried export and check in desktop Gramps, but without succes (Gramps XML option returns empty index.html file)

Does that also trigger a 500 in the backend perhaps?

I suspect somehow the event references of the family contain a corrupted entry.

Can you open F0021 and use the Chrome dev tools to inspect the response for the GET to `/families/F0021`?

**Comment by kotjarka** [comment]
My idea when I found Gramps a while ago was to use only web version of Gramps, because of build-in collaboration. 

Now I installed also desktop version of Gramps, copied existing sqlite.db from Gramps Web into created '/home/user/.gramps/grampsdb/<id>'

Then I digged into Gramps and discovered tool 'repair db'. It repaired some errors.
I also discovered couple of empty events (without relations to persons). I deleted them. I think one of this events might be the cause of error 500.
Then I deleted family F0021 and recreated it again with new ID. I discovered that children person in this family was in db two times by mistake and in relation with family was wrong copy of this person. That was cause of family was not visible on the /tree.
At the end I copied back gramps.db to Gramps Web. Now server error 500 gone and tree is also correct. 

I think that I need to change our workflow and rather sync our home Gramps folders than use Gramps Web, or use it only to view or maybe edit data. I think Gramps Web has a lot of potential but now rather is not ready to standalone using as I planned.

Thanks for your response and Gramps :)

**Comment by DavidMStraub** [comment]
> Then I digged into Gramps and discovered tool 'repair db'. It repaired some errors.
I also discovered couple of empty events (without relations to persons). I deleted them. I think one of this events might be the cause of error 500.

Yes. Gramps does not use a relational database, so a lot of consistency is only implemented at application and not at database level, which can lead to corruption, e.g. by imperfect addons. This is why the Gramps Web documentation describes importing a family tree as a Gramps XML file rather than by directly mounting the databse; this is because exporting and reimporting a Gramps XML file make sure the database is in a consistent state. Performing a database repair is roughly equivalent.

> I think that I need to change our workflow and rather sync our home Gramps folders than use Gramps Web, or use it only to view or maybe edit data. I think Gramps Web has a lot of potential but now rather is not ready to standalone using as I planned.

To be clear, it is not realistics to expect that Gramps Web will ever reach feature parity with Gramps desktop. And I don't think that's desirable. It's not a good use of development resources to maintain to applications that achieve the same thing. I see it as a strength that with Gramps Desktop + Gramps Web we have a powerful pair of open source tools that excel at their specific tasks and are designed to work together.
